#233 tile map extensions for libktx

Quillraven opened the pull request:
First draft of the extensions that I am using in my projects. I had to generalize them a little bit and also tried to follow the kotlin "standard" with naming of some functions like e.g. "propertyOrNull".

Let me know what you think :)

Looking forward to your feedback!

czyzby commented on tiled/README.md:
````diff
@@ -0,0 +1,127 @@
+`MapObject`:
+
+```kotlin
+// assume mapObj refers to a MapObject instance
````
You can make it explicit without an object by starting the example with e.g. `val mapObj: MapObject = getMapObject()`.

czyzby commented on tiled/README.md:
```diff
@@ -0,0 +1,127 @@
+val myOtherProp = mapObj.properties.get("myOtherProperty", Int::class.java)
+val x = mapObj.properties.get("x", 0f, Float::class.java)
+val y = mapObj.properties.get("y", 0f, Float::class.java)
+val shape = (mapObj as RectangleMapObject).rectangle
```
We generally tend to avoid `before`/`after` showcases in the examples section, and instead include in the guide (if necessary). Usage examples should be limited to idiomatic KTX code that you can base your application on.

czyzby commented on tiled/README.md:
```diff
@@ -0,0 +1,127 @@
+val shape = (mapObj as RectangleMapObject).rectangle
+
+// using libktx
+val myProp = mapObj.property("myProperty", "")
```
Explicit type might help here.

czyzby commented on tiled/README.md:
```diff
@@ -0,0 +1,127 @@
+val myProp:String? = map.propertyOrNull("myProp")
+val myProp2 = map.property("myProp2", true)
+val width = map.width
+val totalWidth = map.totalWidth()
```
A lot of these properties seem loosely connected, especially to the iteration example. These could be separate examples with some short introduction, e.g.:

--- 

Iterating over all map objects of a `TiledMap` in a selected layer:

```kotlin
map.forEachMapObject("myLayer") {
    val x = it.x
    val y = it.y
    // create an object at x/y position
}
```

czyzby commented on tiled/README.md:
````diff
@@ -0,0 +1,127 @@
+
+val myLayerProp: Int? = map.layer("myLayer").propertyOrNull("myLayerProp")
+```
+
````
Please include necessary imports in the examples. E.g. `import ktx.tiled.*`. The examples should work (or be easy to fix) after copy-pasting the code.

czyzby commented on tiled/gradle.properties:
```diff
@@ -0,0 +1,2 @@
+projectName=ktx-tiled
+projectDesc=TiledMap utilities for LibGDX applications written with Kotlin.
```
Maybe just `Tiled utilities`?

czyzby commented on tiled/src/main/kotlin/ktx/tiled/mapLayer.kt:
```diff
@@ -0,0 +1,28 @@
+ * @param key property name
+ * @return true if the property exists. Otherwise false
+ */
+fun MapLayer.containsProperty(key: String) = properties.containsKey(key)
```
Missing new line at the end of the file.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/mapLayer.kt:
```diff
@@ -0,0 +1,28 @@
+ * @param key property name
+ * @return true if the property exists. Otherwise false
+ */
+fun MapLayer.containsProperty(key: String) = properties.containsKey(key)
```
This function could be the contains operator, so that `in` would work with a `MapLayer`. Not sure if we want to use it for properties though.

czyzby replied:
For the sake of consistency, we can leave it as `containsProperty` and use `operator contains` for layers in tiled map.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/mapObject.kt:
```diff
@@ -0,0 +1,73 @@
+ * @param defaultValue default value in case the property is missing
+ * @return value of the property or defaultValue if property is missing
+ */
+inline fun <reified T> MapObject.property(key: String, defaultValue: T): T = this.properties[key, defaultValue, T::class.java]
```
Do `MapObject`, `MapLayer` etc. follow any sort of class hierarchy? Maybe the utilities do not have to be added 1 by 1.

Quillraven replied:
Unfortunately no. Those are three separate classes that contain a `MapProperties` attribute which is the container for properties.
The OOTB way to access those properties is by calling `.getProperties()` and then call `get`, `contains`, etc..

That's why I have to add them separately.

But maybe it would be a better approach to make them `MapProperties` extensions instead? Then it would be only once and instead of `property` or `containsProperty` you need to write `properties` or `properties.contains`.

What do you think?

czyzby replied:
I would definitely add extensions for `MapProperties`. If you include the `operator` keyword in the `contains` extension method, it will also allow for the `"key" in map.properties` syntax.

As for the individual extensions to extract properties from `MapObject` and so on, I still think they are useful and we can leave them as they are. In other words, all of them could support the following methods:
- `property(key: String): T`: returns a property or throws an exception. Checks for nulls.
- `property(key: String, defaultValue: T): T`: returns a property or the default value. Checks for nulls.
- `propertyOrNull(key: String): T?`: returns a property or null if undefined.
- `containsProperty(key: String): Boolean`: true if property exists and is not null, false otherwise.

I'd leave `contains` without any suffix for checks specific to the map class. E.g. `TiledMap` could use operator `contains(layerName: String): Boolean` to verify if certain layers exist.

Quillraven replied:
I tried to add an extension now for MapProperties like:

```kotlin
inline operator fun <reified T> MapProperties.get(key: String): T = get(key, T::class.java)
```

but it is shadowed and therefore ignored. So I think it is not possible to add `MapProperties` extensions?

So for now I will stick to TiledMap, MapObject and MapLayer extensions and will provide the functionality discussed above

czyzby replied:
It would be shadowed only if there is a single parameter `get` method already.

Yeah, well, if there's no way to do it, stick to other extensions. Thanks for checking it out.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/mapObject.kt:
```diff
@@ -0,0 +1,73 @@
+ * a x-coordinate then 0 is returned
+ */
+val MapObject.x: Float
+    get() = property("x", 0f)
```
Is this property always defined? An unexpected 0 without any errors or warnings could be misleading. I'd consider converting it to `propertyOrNull("x") ?: throw MissingPropertyException("x")`.

Now that I think about it, `property(name: String)` extension method that return a non-null value or throws an exception would also be a useful utility.

Quillraven replied:
From what I can tell from my maps at least and from the `com.badlogic.gdx.maps.tiled.BaseTmxMapLoader#loadObject` there are some default properties for every object.
Those are `x`, `y`, `width`, `height` and `id` (I did not know about width and height. Just noticed that now and will add them).

There would be other properties like `rotation` and `type` but they can be `null` and are therefore not added and I would not suggest an extension for them due to that.

That's why those properties should never be missing and the default value will never apply. 

I can add the exception extension if you want but personally - at least in my use cases - I prefer a default value instead of an exception.
E.g. I prefer to have an enemy created at 0,0 instead of not creating him at all. The first scenario gives me the hint that everything is working but something is wrong with the location where he is added. The second scenario could be anything since the enemy does not get created at all.
Also, in Kotlin there are no checked exceptions, correct? So a user would need to pro-actively add the try-catch block because the IDE will not warn him?

Again, I am openminded here. Just let me know what do you think is the better approach. Exception or default value - I will do as you wish ;)

czyzby replied:
Exceptions are easier to track and patch. With enemies randomly created at `0, 0`, you wouldn't be sure if the source of the bug is your spawning code or the map settings. With an exception thrown during map loading, you immediately know that the file is corrupted.

It's true that there are no checked exceptions, but then again - if the map files are correct, these kind of errors are not expected at all. You give the user the choice to handle it as they prefer.

czyzby replied:
>  There would be other properties like rotation and type but they can be null and are therefore not added and I would not suggest an extension for them due to that.

You can add nullable properties for these.

Quillraven replied:
changed the standard properties extensions as well to the version where it throws an exception instead of returning a default value

czyzby commented on tiled/src/main/kotlin/ktx/tiled/mapObject.kt:
```diff
@@ -0,0 +1,73 @@
+        is PolylineMapObject -> this.polyline
+        is PolygonMapObject -> this.polygon
+        is RectangleMapObject -> this.rectangle
+        else -> throw UnsupportedOperationException("Shape extension function is not supported for MapObject of type ${this::class.java}")
```
This exception might be a bit too generic, I'd add something along `MissingPropertyException` in `ktx-tiled` module to handle this and similar cases.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,99 @@
+import com.badlogic.gdx.maps.MapProperties
+import com.badlogic.gdx.maps.tiled.TiledMap
+
+private val DEFAULT_MAP_LAYER = MapLayer().apply { name = "ktx-default-map-layer" }
```
I don't think this is a good approach, especially since `MapLayer` objects seem to be mutable.

Quillraven replied:
You are right about the mutable part. It could be changed from the outside which is not nice.

The reason why I do it like this is because of my use-cases that I had. E.g. when I load a map I parse different layers and do something with them:
```kotlin
// this refers to a TiledMap
createSceneryEntities(this)
createEnemyEntities(this)
createSavepoints(this)
createNPCs(this)
createItemEntities(this)
createPortalEntities(this)
createTriggers(this)
```

One of these methods then looks like this:
```kotlin
private fun createSceneryEntities(map: Map) {
    map.mapObjects(LAYER_COLLISION).forEach { mapObj ->
        // loop through all map objects of that layer and
        // create the scenery entity according to the shape
        // of the map object.
        ecsEngine.scenery(world, mapObj.shape)
    }
}
```

So I don't care if the layer is there or not. If it is there then I will iterate over its objects and create collision scenery box2d objects out of it.
I would not like to have an exception because it is totally valid to leave out the collision layer.
Also, I would not like to have `isLayerExisting` calls everywhere. Granted, with Kotlin it is just a simple `?` character but especially when it comes to chaining I might end up with several `?`.

Of course this was my use-case and we should think here of a more generic approach for everyone. Just wanted to explain to you why I do it that way.

At least for me we can get rid of it since the `forEachMapObject` extension can take care of it.

I will make the change.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,99 @@
+ * @param key property name
+ * @return true if the property exists. Otherwise false
+ */
+fun TiledMap.containsProperty(key: String) = properties.containsKey(key)
```
This could also be an operator.

czyzby replied:
Maybe a `contains` operator would work better to check for existing layers in case of a `TiledMap` though.

Quillraven replied:
refer to `MapProperties` extension proposal from `MapObject`

czyzby commented on tiled/src/main/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,99 @@
+ * a **tileheight** property then 0 is returned
+ */
+val TiledMap.tileHeight: Int
+    get() = property("tileheight", 0)
```
Same comment about defaults applies here. I think I'd rather have to handle an exception rather than get an unexpected 0 on tile height, width, etc.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,99 @@
+ * @param layerName name of [MapLayer]
+ * @return [MapLayer] of given name or an empty default map layer, if the specified layer does not exist
+ */
+fun TiledMap.layer(layerName: String) = layers[layerName] ?: DEFAULT_MAP_LAYER
```
OK, getting a default instance of a `MapLayer` would not be very useful and it could lead to unexpected behaviors. I wouldn't have suspected it as a user. You should definitely throw an exception if the map layer is missing.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,99 @@
+ */
+fun TiledMap.forEachMapObject(layerName: String, action: (MapObject) -> Unit) {
+    layer(layerName).objects.forEach { action(it) }
+}
```
Missing new line.

czyzby commented on tiled/src/test/kotlin/ktx/tiled/mapLayer.kt:
```diff
@@ -0,0 +1,33 @@
+package ktx.tiled
```
Add `Test` suffix to the file name.

czyzby replied:
Same goes for other tests.

czyzby commented on tiled/src/test/kotlin/ktx/tiled/mapLayer.kt:
```diff
@@ -0,0 +1,33 @@
+        Assert.assertTrue(mapLayer.containsProperty("active"))
+        Assert.assertFalse(mapLayer.containsProperty("x"))
+    }
+}
```
Missing new line.

czyzby commented on tiled/src/test/kotlin/ktx/tiled/mapObjects.kt:
```diff
@@ -0,0 +1,75 @@
+    fun `retrieve shape from unsupported MapObject`() {
+        textureObject.shape
+    }
+}
```
Missing new line.

czyzby commented on tiled/src/test/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,84 @@
+        tiledMap.forEachMapObject("non-existing") { ++counter }
+        assertEquals(0, counter)
+    }
+}
```
Missing new line.

Quillraven commented:
I have updated now all the things according to your feedback. 
The examples of the README.md should be also clearer now.

czyzby commented:
@Quillraven I believe you have missed the comments that were automatically wrapped by GitHub. I've resolved some of the comments that were already addressed, please go through the entire review and check out the rest. Thanks!

czyzby commented:
To summarize the most important changes:
* Add a `MissingPropertyException` (and other similar exceptions, if necessary) that extends `GdxRuntimeException`.
* Add `property(name: String): T` extension methods that throw `MissingPropertyException` if the property is missing. Using a default value or handling nulls manually is not always desired.
* Do not return a default, static `MapLayer` instance if a layer with a particular name is missing. This behavior is error-prone.

Quillraven commented:
ah I see it now. Don't know why it didn't show me that before. Most likely I was blind  ;)

Will respond to it later today.

czyzby commented:
No problem. That's why I usually prefer the [changes view](https://github.com/libktx/ktx/pull/234/files), as it's harder to miss the comments.

Quillraven commented:
Added some comments to your remarks and changed a few things.

Main change would now be to make `MapProperties` extensions instead of having `MapObject`, `MapLayer` and `TiledMap` extensions. 

This would lead to changes like instead of `map.property("myProperty", "")` we would write `map.properties("myProperty", "")`. So basically property becomes properties ;)

What do you think?

czyzby commented:
I wouldn't expect a map of properties to be callable. If anything, I think square braces would be a standard syntax for this: `map.properties["myProperty"]`. If `get` currently requires a `Class` instance, we could certainly improve it with an operator `get` extension method.

czyzby commented:
Btw, @Quillraven, you didn't have to remove the `layer` utility altogether, I'd just prefer if it threw an exception in case of an unknown layer rather than return a default empty instance.

```kotlin
/**
 * Extension method to retrieve a non-null [MapLayer] of the [TiledMap]. In case the layer
 * cannot be found, a default empty layer is returned with name **ktx-default-map-layer**.
 * @param layerName name of [MapLayer]
 * @return [MapLayer] of given name or an empty default map layer, if the specified layer does not exist
* @throws UnknownLayerException if the given name is missing.
 */
fun TiledMap.layer(layerName: String) = layers[layerName] ?: throw UnknownLayerException(layerName)
```

Quillraven commented:
I have looked now into all open topics and adjusted the extensions accordingly. Also added the remaining missing standard properties, in case someone needs them.

I was not able to add extensions for `MapProperties` class due to shadowing issues which I don't know how to solve (maybe they are not solvable?).
Therefore, I kept the MapObject, MapLayer and TiledMap extensions.

Also added contains to TiledMap and the layer extension, this time throwing one of the newly added exceptions.

Let me know if this looks good then I will update the README.md accordingly. I did not want to do that now because I want to wait for the final version. Otherwise I will update it multiple times ;)

czyzby commented on tiled/src/main/kotlin/ktx/tiled/exception.kt:
```diff
@@ -0,0 +1,23 @@
+/**
+ * [GdxRuntimeException] that is thrown when trying to access a shape of a [MapObject] that do not have any shape like [TextureMapObject]
+ */
+class MissingShapeException(message: String) : GdxRuntimeException(message)
```
Missing new line.

Quillraven replied:
Out of curiosity, what is the problem if a new line is missing? Seems like auto-formatting does not add it automatically on my side.

czyzby replied:
[Here's a few reasons](https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file). You can configure IntelliJ to [add it automatically](https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file).

czyzby commented on tiled/src/main/kotlin/ktx/tiled/mapLayer.kt:
```diff
@@ -0,0 +1,38 @@
+ * @param defaultValue default value in case the property is missing
+ * @return value of the property or defaultValue if property is missing
+ */
+inline fun <reified T> MapLayer.property(key: String, defaultValue: T): T = this.properties[key, defaultValue, T::class.java]
```
Can this call return a null? E.g. if the key is linked to an existing property with `null` value. It would make more sense to throw MissingPropertyException than receive a NPE. The same applies to other such calls.

Quillraven replied:
In my opinion it cannot do that and it is also part of the Tests that I wrote. When checking the original method, from what I can tell, it will always return the defaultValue if either the key is missing or the linked value to the key is null.

```Java
public <T> T get (String key, T defaultValue, Class<T> clazz) {
	Object object = get(key);
	return object == null ? defaultValue : (T)object;
}
```

czyzby replied:
OK, thanks for verifitcation.

Quillraven commented:
Updated ReadMe now and also added the property extensions to Tile and TileSet. Those should be all objects that use the `MapProperties` class.

czyzby commented on tiled/src/main/kotlin/ktx/tiled/tiledMap.kt:
```diff
@@ -0,0 +1,152 @@
+ * @param layerName name of [MapLayer]
+ * @param action action to execute per [MapObject] of the [MapLayer]
+ */
+fun TiledMap.forEachMapObject(layerName: String, action: (MapObject) -> Unit) {
```
This function can be inlined.

czyzby commented:
Looks OK now. I might do some minor documentation changes here and there, but I'm merging this module as is. Thank you!

czyzby commented:
@Quillraven Just a small tip on the tests. You don't have to store every object as a class property, especially if it's relevant only for a single test. It's OK to create objects inside the test functions.

I'm also a fan of the [given-when-then](https://en.wikipedia.org/wiki/Given-When-Then) testing structure. While Kotlin frameworks do not support this as keywords as e.g. Groovy's Spock does, it makes sense to compose your tests like that.

Basically, you divide your tests into 3 sections:
- given: preparation. You create and set up all necessary object here.
- when: execution. Ideally, a single tested method call.
- then: assertion. All of the checks that you need to perform.

You'll notice that a lot of the new and refactored tests in KTX are written like that.

Quillraven commented:
Ah I know about that because we do something similar for typescript tests at work.
Was not aware that this is actually a good practice. Will improve on that the nex time, thanks.

About the object creation... in this case I wanted to have it "global" because I wanted to reuse it amongst all tests but maybe that is not ideal. 

Thank you for your explanations and help. I learn from that  ;)
